### PR TITLE
[FLINK-26554] Upgrade Operator SDK to avoid cleanup race condition

### PR DESCRIPTION
**Pull request by gyfora**
Upgrade the SDK version thus eliminating some annoying exception logs on cleanup caused by a race condition.

### CODE REVIEW

**Comment by gyfora** [comment]
cc @tweise

**Review by tweise — APPROVED** [review]
Nice!

**Comment by gyfora** [comment]
The e2e teat failures are strange, I specifically added the boolean flag to disable an unnecessery check that causes these errors. Works well locally. Have to check this tomorrow

**Comment by gyfora** [comment]
@wangyang0918 @mbalassi any idea why this might happen?

**Comment by wangyang0918** [comment]
```
2022-03-21 18:17:17,693 i.j.o.a.c.ExecutorServiceManager [DEBUG] Initialized ExecutorServiceManager executor: class java.util.concurrent.ThreadPoolExecutor, timeout: 10
Error: Exception in thread "main" io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.96.0.1/apis/apiextensions.k8s.io/v1/customresourcedefinitions/flinkdeployments.flink.apache.org. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. customresourcedefinitions.apiextensions.k8s.io "flinkdeployments.flink.apache.org" is forbidden: User "system:serviceaccount:default:flink-operator" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope.
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:682)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:661)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:610)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:555)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:[518](https://github.com/apache/flink-kubernetes-operator/runs/5632398750?check_suite_focus=true#step:9:518))
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleGet(OperationSupport.java:487)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleGet(OperationSupport.java:457)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleGet(BaseOperation.java:698)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.getMandatory(BaseOperation.java:184)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:151)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:83)
	at io.javaoperatorsdk.operator.processing.Controller.start(Controller.java:178)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(Unknown Source)
	at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(Unknown Source)
	at java.base/java.util.concurrent.CountedCompleter.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doInvoke(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(Unknown Source)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(Unknown Source)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.forEach(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(Unknown Source)
	at io.javaoperatorsdk.operator.Operator$ControllerManager.start(Operator.java:164)
	at io.javaoperatorsdk.operator.Operator.start(Operator.java:82)
	at org.apache.flink.kubernetes.operator.FlinkOperator.main(FlinkOperator.java:84)
```
@gyfora Maybe we could add the following codes in `e2e-tests/utils.sh#debug_and_show_logs` to print the actual RBAC.

```
    kubectl get serviceaccount flink-operator -o yaml
    kubectl get clusterrole flink-operator -o yaml
    kubectl get clusterrolebinding flink-operator-cluster-role-binding -o yaml
```

**Comment by wangyang0918** [comment]
@gyfora I think the root cause is https://github.com/java-operator-sdk/java-operator-sdk/pull/1016. The `checkCRDAndValidateLocalModel` never works because of this bug.

File: helm/flink-operator/templates/flink-operator.yaml [hunk 1]
```diff
@@ -67,6 +67,8 @@ spec:
               value: -Dlog4j.configurationFile=/opt/flink-operator/conf/log4j2.properties
             - name: FLINK_OPERATOR_WATCH_NAMESPACES
               value: {{ join "," .Values.watchNamespaces  }}
+            - name: JAVA_OPERATOR_SDK_CHECK_CRD
```
Comment (wangyang0918) [inline helm/flink-operator/templates/flink-operator.yaml:70]: Then ENV doest not work. They only support java properties now.

https://github.com/java-operator-sdk/java-operator-sdk/pull/1016/files

**Review by wangyang0918 — CHANGES_REQUESTED** [review]

**Comment by gyfora** [comment]
@wangyang0918 , I talked the operator sdk guys yesterday on Discord. They introduced this check and made it default (by accident) in `2.1.2`. 

This check is intended to be off for production deployments.

For me with these changes it works correctly locally so I dont understand why it breaks on github so I need to debug further

**Comment by wangyang0918** [comment]
@gyfora It seems that the e2e tests are not stable after #84.

```
Run ls e2e-tests/test_*.sh | while read script_test;do \
Running e2e-tests/test_kubernetes_application_ha.sh
persistentvolumeclaim/flink-example-statemachine created
Error from server (InternalError): error when creating "e2e-tests/data/cr.yaml": Internal error occurred: failed calling webhook "vflinkdeployments.flink.apache.org": failed to call webhook: Post "https://flink-operator-webhook-service.default.svc:443/validate?timeout=10s": dial tcp 10.106.63.26:443: connect: connection refused
Command: kubectl apply -f e2e-tests/data/cr.yaml failed. Retrying...
flinkdeployment.flink.apache.org/flink-example-statemachine created
persistentvolumeclaim/flink-example-statemachine unchanged
Error from server (NotFound): deployments.apps "flink-example-statemachine" not found
Command: kubectl get deploy/flink-example-statemachine failed. Retrying...
NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
flink-example-statemachine   0/1     1            0           1s
deployment.apps/flink-example-statemachine condition met
Waiting for jobmanager pod flink-example-statemachine-7fcf55c88b-h5r7r ready.
pod/flink-example-statemachine-7fcf55c88b-h5r7r condition met
Waiting for log "Rest endpoint listening at"...
Log "Rest endpoint listening at" shows up.
Waiting for log "Completed checkpoint [0-[9](https://github.com/apache/flink-kubernetes-operator/runs/5640468148?check_suite_focus=true#step:9:9)]+ for job"...
Log "Completed checkpoint [0-9]+ for job" shows up.
Successfully verified that flinkdep/flink-example-statemachine.status.jobManagerDeploymentStatus is in READY state.
Successfully verified that flinkdep/flink-example-statemachine.status.jobStatus.state is in RUNNING state.
Kill the flink-example-statemachine-7fcf55c88b-h5r7r
Defaulted container "flink-main-container" out of: flink-main-container, artifacts-fetcher (init)
Waiting for log "Restoring job 00000000000000000000000000000000 from Checkpoint"...
Log "Restoring job 00000000000000000000000000000000 from Checkpoint" shows up.
Waiting for log "Completed checkpoint [0-9]+ for job"...
Log "Completed checkpoint [0-9]+ for job" shows up.
Status verification for flinkdep/flink-example-statemachine.status.jobManagerDeploymentStatus failed. It is DEPLOYED_NOT_READY instead of READY.
Debugging failed e2e test:
```

**Review by wangyang0918 — APPROVED** [review]
This PR should work now. +1 for merging.

**Comment by Aitozi** [comment]
> @gyfora It seems that the e2e tests are not stable after #84.
> 
> ```
> Run ls e2e-tests/test_*.sh | while read script_test;do \
> Running e2e-tests/test_kubernetes_application_ha.sh
> persistentvolumeclaim/flink-example-statemachine created
> Error from server (InternalError): error when creating "e2e-tests/data/cr.yaml": Internal error occurred: failed calling webhook "vflinkdeployments.flink.apache.org": failed to call webhook: Post "https://flink-operator-webhook-service.default.svc:443/validate?timeout=10s": dial tcp 10.106.63.26:443: connect: connection refused
> Command: kubectl apply -f e2e-tests/data/cr.yaml failed. Retrying...
> flinkdeployment.flink.apache.org/flink-example-statemachine created
> persistentvolumeclaim/flink-example-statemachine unchanged
> Error from server (NotFound): deployments.apps "flink-example-statemachine" not found
> Command: kubectl get deploy/flink-example-statemachine failed. Retrying...
> NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
> flink-example-statemachine   0/1     1            0           1s
> deployment.apps/flink-example-statemachine condition met
> Waiting for jobmanager pod flink-example-statemachine-7fcf55c88b-h5r7r ready.
> pod/flink-example-statemachine-7fcf55c88b-h5r7r condition met
> Waiting for log "Rest endpoint listening at"...
> Log "Rest endpoint listening at" shows up.
> Waiting for log "Completed checkpoint [0-[9](https://github.com/apache/flink-kubernetes-operator/runs/5640468148?check_suite_focus=true#step:9:9)]+ for job"...
> Log "Completed checkpoint [0-9]+ for job" shows up.
> Successfully verified that flinkdep/flink-example-statemachine.status.jobManagerDeploymentStatus is in READY state.
> Successfully verified that flinkdep/flink-example-statemachine.status.jobStatus.state is in RUNNING state.
> Kill the flink-example-statemachine-7fcf55c88b-h5r7r
> Defaulted container "flink-main-container" out of: flink-main-container, artifacts-fetcher (init)
> Waiting for log "Restoring job 00000000000000000000000000000000 from Checkpoint"...
> Log "Restoring job 00000000000000000000000000000000 from Checkpoint" shows up.
> Waiting for log "Completed checkpoint [0-9]+ for job"...
> Log "Completed checkpoint [0-9]+ for job" shows up.
> Status verification for flinkdep/flink-example-statemachine.status.jobManagerDeploymentStatus failed. It is DEPLOYED_NOT_READY instead of READY.
> Debugging failed e2e test:
> ```

> Status verification for flinkdep/flink-example-statemachine.status.jobManagerDeploymentStatus failed. It is DEPLOYED_NOT_READY instead of READY.

I also encountered this same error today on my private CI, re-run the job, it passed.

**Comment by mbalassi** [comment]
Opening a ticket for this, I run into it too. I have an idea for resolving it.

> > @gyfora It seems that the e2e tests are not stable after #84.
> > ```
> > Run ls e2e-tests/test_*.sh | while read script_test;do \
> > Running e2e-tests/test_kubernetes_application_ha.sh
> > persistentvolumeclaim/flink-example-statemachine created
> > Error from server (InternalError): error when creating "e2e-tests/data/cr.yaml": Internal error occurred: failed calling webhook "vflinkdeployments.flink.apache.org": failed to call webhook: Post "https://flink-operator-webhook-service.default.svc:443/validate?timeout=10s": dial tcp 10.106.63.26:443: connect: connection refused
> > Command: kubectl apply -f e2e-tests/data/cr.yaml failed. Retrying...
> > flinkdeployment.flink.apache.org/flink-example-statemachine created
> > persistentvolumeclaim/flink-example-statemachine unchanged
> > Error from server (NotFound): deployments.apps "flink-example-statemachine" not found
> > Command: kubectl get deploy/flink-example-statemachine failed. Retrying...
> > NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
> > flink-example-statemachine   0/1     1            0           1s
> > deployment.apps/flink-example-statemachine condition met
> > Waiting for jobmanager pod flink-example-statemachine-7fcf55c88b-h5r7r ready.
> > pod/flink-example-statemachine-7fcf55c88b-h5r7r condition met
> > Waiting for log "Rest endpoint listening at"...
> > Log "Rest endpoint listening at" shows up.
> > Waiting for log "Completed checkpoint [0-[9](https://github.com/apache/flink-kubernetes-operator/runs/5640468148?check_suite_focus=true#step:9:9)]+ for job"...
> > Log "Completed checkpoint [0-9]+ for job" shows up.
> > Successfully verified that flinkdep/flink-example-statemachine.status.jobManagerDeploymentStatus is in READY state.
> > Successfully verified that flinkdep/flink-example-statemachine.status.jobStatus.state is in RUNNING state.
> > Kill the flink-example-statemachine-7fcf55c88b-h5r7r
> > Defaulted container "flink-main-container" out of: flink-main-container, artifacts-fetcher (init)
> > Waiting for log "Restoring job 00000000000000000000000000000000 from Checkpoint"...
> > Log "Restoring job 00000000000000000000000000000000 from Checkpoint" shows up.
> > Waiting for log "Completed checkpoint [0-9]+ for job"...
> > Log "Completed checkpoint [0-9]+ for job" shows up.
> > Status verification for flinkdep/flink-example-statemachine.status.jobManagerDeploymentStatus failed. It is DEPLOYED_NOT_READY instead of READY.
> > Debugging failed e2e test:
> > ```
> 
> > Status verification for flinkdep/flink-example-statemachine.status.jobManagerDeploymentStatus failed. It is DEPLOYED_NOT_READY instead of READY.
> 
> I also encountered this same error today on my private CI, re-run the job, it passed.

**Comment by SteNicholas** [comment]
@gyfora, the Operator SDK is upgraded to [2.1.3](https://github.com/java-operator-sdk/java-operator-sdk/releases/tag/v2.1.3). Should we bump to the release version?

**Pull request merged**